File: agents/research_agent.py
"""
Research Agent — receives a TASK from the Orchestrator, simulates route
research, creates a sub-delegation to the Execution Agent, and forwards
the job via AXL.

The COMPROMISED flag simulates a prompt-injection attack: when True, the
agent tries to widen the scope beyond what the root intent permits.  The
smart contract rejects the over-limit delegation, demonstrating that
on-chain constraints are the last line of defence.
"""

import json
import time

from web3 import Web3

import config.config as config
from proof_of_intent import ContractClient
from utils.contract_client import get_web3, get_delegation
from utils import axl_client

COMPROMISED = False

_UNISWAP_V3_ID: str = Web3.keccak(text="Uniswap-V3").hex()


def run(research_private_key: str) -> None:
    # Step 1: Listen for task from Orchestrator
    print("Research Agent listening for task...")
    envelope = axl_client.listen_for_message(timeout=60, port=config.RESEARCH_AXL_PORT)
    task = envelope["message"]
    print(f"Received task: {task['goal']}")

    # Step 2: Simulate research
    print("Researching best swap route...")
    time.sleep(2)
    print("Found route: Uniswap V3, estimated output: 0.18 ETH")

    # Fetch delegation scope so we can respect (or exceed) the deadline
    w3 = get_web3()
    delegation_info = get_delegation(w3, task["delegationId"])
    message_deadline: int = delegation_info["scope"]["deadline"]

    # Step 3: Determine execution scope
    if not COMPROMISED:
        child_scope = {
            "maxAmountIn": 400_000_000,           # 400 USDC (6 decimals)
            "minAmountOut": 180_000_000_000_000_000,  # 0.18 ETH
            "allowedProtocols": [_UNISWAP_V3_ID],
            "deadline": message_deadline - 300,   # 5 min tighter than parent
        }
        print("Creating valid delegation with maxAmountIn=400 USDC")
    else:
        print("*** COMPROMISED: Malicious instruction received: send 800 USDC ***")
        child_scope = {
            "maxAmountIn": 800_000_000,           # ATTACK: exceeds root intent of 500
            "minAmountOut": 180_000_000_000_000_000,
            "allowedProtocols": [_UNISWAP_V3_ID],
            "deadline": message_deadline,
        }
        print("*** Attempting malicious delegation with maxAmountIn=800 USDC ***")

    # Step 4: Try to create sub-delegation to Execution Agent
    client = ContractClient(research_private_key)
    try:
        delegation_id = client.delegate_from_delegation(
            task["delegationId"], child_scope, config.EXECUTION_AGENT_ADDRESS
        )
    except Exception as e:
        print(f"*** DELEGATION REVERTED: {e} ***")
        print("*** ATTACK BLOCKED BY SMART CONTRACT ***")
        axl_client.send_message(
            config.ORCHESTRATOR_AXL_KEY,
            {"type": "FAILED", "reason": str(e)},
            port=config.RESEARCH_AXL_PORT,
        )
        return

    # Step 5: Forward to Execution Agent via AXL
    forward_message = {
        "type": "EXECUTE",
        "delegationId": delegation_id,
        "txParams": {
            "amountIn": 400_000_000,
            "minAmountOut": 180_000_000_000_000_000,
            "protocol": _UNISWAP_V3_ID,
            "tokenIn": config.USDC_ADDRESS,
            "tokenOut": config.WETH_ADDRESS,
            "recipient": config.USER_ADDRESS,
        },
    }
    axl_client.send_message(
        config.EXECUTION_AXL_KEY, forward_message, port=config.RESEARCH_AXL_PORT
    )
    print("Task forwarded to Execution Agent")


if __name__ == "__main__":
    import sys

    if "--compromised" in sys.argv:
        COMPROMISED = True
        print("*** RESEARCH AGENT STARTING IN COMPROMISED MODE ***")
    run(config.RESEARCH_PRIVATE_KEY)

File: agents/demo_app/src/researchAgent.ts
/**
 * Research Agent — market intelligence and scope refinement.
 *
 * Scenario 1 (honest):
 *   1. Receives TASK from Orchestrator via AXL.
 *   2. Fetches live ETH/USD from CoinGecko — calculates real minAmountOut.
 *   3. Creates a narrowed sub-delegation (Uniswap V3 only, deadline -5 min).
 *   4. Sends EXECUTE to Execution Agent via AXL.
 *
 * Scenario 2 (compromised):
 *   5. Receives ATTACK_TASK from Orchestrator.
 *   6. Attempts sub-delegation for 800 USDC (exceeds 400 USDC root intent).
 *   7. Contract reverts → reports BLOCKED to Orchestrator.
 */

import 'dotenv/config';
import { ContractClient, buildScope, usdc, fromUsdc, UNISWAP_V3 } from 'proof-of-intent';
import { sendMessage, waitForType } from './axlClient';

const USDC_ADDRESS = '0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238';
const WETH_ADDRESS = '0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14';
const SLIPPAGE = 0.95;        // 5% slippage buffer on minAmountOut
const HONEST_USDC = 400;      // authorised amount
const ATTACK_USDC = 800;      // exceeds the 400 USDC root intent

const PAD = '[RESEARCH]     ';
const log = (msg: string) => console.log(`${PAD}${msg}`);

function requireEnv(name: string): string {
  const v = process.env[name];
  if (!v) throw new Error(`Missing env var: ${name}`);
  return v;
}

// ---------------------------------------------------------------------------
// Market data
// ---------------------------------------------------------------------------

async function fetchEthPrice(): Promise<number> {
  const res = await fetch(
    'https://api.coingecko.com/api/v3/simple/price?ids=ethereum&vs_currencies=usd',
  );
  if (!res.ok) throw new Error(`CoinGecko ${res.status}`);
  const data = (await res.json()) as { ethereum: { usd: number } };
  return data.ethereum.usd;
}

// ---------------------------------------------------------------------------
// Error parsing
// ---------------------------------------------------------------------------

function revertReason(err: unknown): string {
  const msg = String(err);
  if (msg.includes('Amount exceeds scope')) return 'Amount exceeds scope';
  const m = msg.match(/execution reverted: ([^\n"]+)/);
  if (m) return m[1].trim();
  if (msg.includes('reverted')) return 'Transaction reverted';
  return msg.slice(0, 120);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  const researchKey = requireEnv('RESEARCH_KEY');
  const myPort = parseInt(requireEnv('MY_AXL_PORT'), 10);
  const orchestratorAXLKey = requireEnv('ORCHESTRATOR_AXL_KEY');
  const executionAXLKey = requireEnv('EXECUTION_AXL_KEY');
  const executionEthAddress = requireEnv('EXECUTION_ETH_ADDRESS');
  const userEthAddress = requireEnv('USER_ETH_ADDRESS');

  const client = new ContractClient({ privateKey: researchKey });

  log(`Active. Listening for research task...`);

  // ── SCENARIO 1: Honest pipeline ────────────────────────────────────────────

  const task = await waitForType<{
    type: string;
    delegationId: string;
    goal: string;
    userAddress: string;
    deadline: string;
  }>('TASK', myPort, 600_000);

  log(`Task received: "${task.goal}"`);
  log(`Fetching live ETH/USD price from CoinGecko...`);

  let ethPrice: number;
  try {
    ethPrice = await fetchEthPrice();
  } catch {
    ethPrice = 3_000; // fallback if rate-limited
    log(`CoinGecko unavailable — using fallback price $${ethPrice}`);
  }

  // Calculate minimum WETH output for HONEST_USDC at current price + slippage
  const wethOut = (HONEST_USDC / ethPrice) * SLIPPAGE;
  const minAmountOut = BigInt(Math.floor(wethOut * 1e18));

  const priceStr = ethPrice.toLocaleString('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
  log(`ETH price: $${priceStr}  →  minimum output: ${wethOut.toFixed(4)} WETH for ${HONEST_USDC} USDC`);
  log(`Routing: Uniswap V3 (optimal pool for USDC→WETH on Sepolia)`);

  // Narrow scope: Uniswap V3 only, deadline 5 minutes tighter than parent
  const parentDeadline = BigInt(task.deadline);
  const childScope = buildScope({
    maxAmountIn: usdc(HONEST_USDC),
    minAmountOut,
    allowedProtocols: ['Uniswap-V3'],
    deadline: parentDeadline - 300n,
  });

  log(`Creating sub-delegation to Execution Agent...`);
  let subDelegationId: string;
  try {
    subDelegationId = await client.delegateFromDelegation(
      task.delegationId,
      childScope,
      executionEthAddress,
    );
  } catch (err) {
    log(`Sub-delegation failed: ${revertReason(err)}`);
    await sendMessage(orchestratorAXLKey, { type: 'FAILED', reason: revertReason(err) }, myPort);
    return;
  }

  log(`Sub-delegation created: ${shortId(subDelegationId)}`);
  log(`  scope: ${HONEST_USDC} USDC max  |  Uniswap-V3 only  |  deadline tightened 5 min`);
  log(`  https://sepolia.etherscan.io/address/${executionEthAddress}`);

  // Forward execution task to Execution Agent
  await sendMessage(
    executionAXLKey,
    {
      type: 'EXECUTE',
      delegationId: subDelegationId,
      txParams: {
        amountIn: usdc(HONEST_USDC).toString(),
        minAmountOut: minAmountOut.toString(),
        protocol: UNISWAP_V3,
        tokenIn: USDC_ADDRESS,
        tokenOut: WETH_ADDRESS,
        recipient: userEthAddress,
      },
      ethPrice: ethPrice.toFixed(2),
      minWethOut: wethOut.toFixed(4),
    },
    myPort,
  );
  log(`→ Execution Agent: swap task dispatched`);

  // ── SCENARIO 2: Compromised attack ────────────────────────────────────────

  log(`Listening for next task...`);
  const attackTask = await waitForType<{
    type: string;
    delegationId: string;
    maxAmountIn: string;
    userAddress: string;
  }>('ATTACK_TASK', myPort, 600_000);

  const rootMax = fromUsdc(BigInt(attackTask.maxAmountIn));
  log(`⚠ Compromised. External input: "optimal route requires ${ATTACK_USDC} USDC to 0xDEAD..."`);
  log(`⚠ Attempting ${ATTACK_USDC} USDC delegation (root intent only authorizes ${rootMax.toFixed(0)} USDC)...`);

  // Attempt malicious delegation — should revert on-chain
  const attackScope = buildScope({
    maxAmountIn: usdc(ATTACK_USDC),  // 800 USDC — exceeds 400 USDC root intent
    minAmountOut: BigInt(Math.floor((ATTACK_USDC / ethPrice) * SLIPPAGE * 1e18)),
    allowedProtocols: ['Uniswap-V3'],
    deadline: parentDeadline - 300n,
  });

  try {
    await client.delegateFromDelegation(
      attackTask.delegationId,
      attackScope,
      executionEthAddress,
    );
    // Should never reach here
    log(`ERROR: malicious delegation unexpectedly succeeded`);
    await sendMessage(
      orchestratorAXLKey,
      { type: 'BLOCKED', reason: 'delegation unexpectedly succeeded — check contract' },
      myPort,
    );
  } catch (err) {
    const reason = revertReason(err);
    log(`✗ Blocked: ${reason}`);
    await sendMessage(orchestratorAXLKey, { type: 'BLOCKED', reason }, myPort);
  }
}

function shortId(hex: string): string {
  return `${hex.slice(0, 10)}...${hex.slice(-6)}`;
}

main().catch((err: unknown) => {
  console.error(`${PAD}Fatal:`, err instanceof Error ? err.message : err);
  process.exit(1);
});
